hack/ci/setup-kind-cluster.sh

```diff
@@ -0,0 +1,232 @@
+EOF
+
+  kind create cluster --config kind-config.yaml
+  pushElapsed kind_cluster_create_duration_milliseconds $beforeKindCreate
+
+  # unwrap the socket inside the kind cluster and make it available on a TCP port,
+  # because containerd/Docker doesn't support sockets for mirrors.
+  docker exec $KIND_CLUSTER_NAME-control-plane bash -c 'socat TCP4-LISTEN:5001,fork,reuseaddr UNIX:/mirror/mirror.sock &'
+else
+  kind create cluster --config kind-config.yaml
+fi
+
+echodate "Kind cluster $KIND_CLUSTER_NAME is up and running."
+
+if [ ! -f cni-plugin-deployed ]; then
+  echodate "Installing CNI plugin."
+  (
+    # Install CNI plugins since they are not installed by default in KIND. Also, kube-flannel doesn't install
+    # CNI plugins unlike other plugins so we have to do it manually.
+    setup_cni_in_kind=$(cat hack/ci/setup-cni-in-kind.sh)
+    docker exec $KIND_CLUSTER_NAME-control-plane bash -c "$setup_cni_in_kind &"
+  )
+
+  kubectl create -f https://raw.githubusercontent.com/flannel-io/flannel/v0.18.0/Documentation/kube-flannel.yml
+  touch cni-plugin-deployed
+fi
+
+if [ -z "${DISABLE_CLUSTER_EXPOSER:-}" ]; then
+  # Annotate kube-apiserver service so that the cluster exposer can expose it
+  kubectl annotate svc kubernetes -n default nodeport-proxy.k8s.io/expose=true
+
+  # Start cluster exposer, which will expose services from within kind as
+  # a NodePort service on the host
+  echodate "Starting cluster exposer"
+  (
+    # Clone kubermatic repo to build clusterexposer
+    mkdir -p /tmp/kubermatic
+    cd /tmp/kubermatic
+    echodate "Cloning cluster exposer"
+    KKP_REPO_URL="${KKP_REPO_URL:-https://github.com/kubermatic/kubermatic.git}"
+    KKP_REPO_TAG="${KKP_REPO_BRANCH:-master}"
+    git clone --depth 1 --branch "${KKP_REPO_TAG}" "${KKP_REPO_URL}" .
+
+    echodate "Building cluster exposer"
+    CGO_ENABLED=0 go build --tags ce -v -o /tmp/clusterexposer ./pkg/test/clusterexposer/cmd
+  )
+
+  export KUBECONFIG=~/.kube/config
+  /tmp/clusterexposer \
+    --kubeconfig-inner "$KUBECONFIG" \
+    --kubeconfig-outer "/etc/kubeconfig/kubeconfig" \
+    --build-id "$PROW_JOB_ID" &> /var/log/clusterexposer.log &
+
+  function print_cluster_exposer_logs {
+    if [[ $? -ne 0 ]]; then
+      # Tolerate errors and just continue
+      set +e
+      echodate "Printing cluster exposer logs"
+      cat /var/log/clusterexposer.log
+      echodate "Done printing cluster exposer logs"
+      set -e
+    fi
+  }
+  appendTrap print_cluster_exposer_logs EXIT
+
+  TEST_NAME="Wait for cluster exposer"
+  echodate "Waiting for cluster exposer to be running"
+
+  retry 5 curl -s --fail http://127.0.0.1:2047/metrics -o /dev/null
+  echodate "Cluster exposer is running"
+
+  echodate "Setting up iptables rules to make nodeports available"
+  KIND_NETWORK_IF=$(ip -br addr | grep -- 'br-' | cut -d' ' -f1)
+
+  iptables -t nat -A PREROUTING -i eth0 -p tcp -m multiport --dports=30000:33000 -j DNAT --to-destination 172.18.0.2
```

Comment:
It may be a bit dangerous to hardcode the IP `172.18.0.2`. For example, if the node would have a different docker container already running before, it could get `172.18.0.3`, or the node can be configured to use a different docker subnet (e.g. after docker version upgrade). Some kind of `docker inspect` to determine the actual IP would be safer.

Reply:
Updated